### PR TITLE
Fix CompositeBytesReference#slice to not throw AIOOBE with legal offsets.

### LINKED ISSUE #35950: [CI] CompositeBytesReferenceTests::testSlice failure
seen in https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+intake/316/console

repros with:
```
./gradlew :server:test -Dtests.seed=24CCFF659D275216 -Dtests.class=org.elasticsearch.common.bytes.CompositeBytesReferenceTests -Dtests.method="testSlice" -Dtests.security.manager=true -Dtests.locale=es-CR -Dtests.timezone=America/Chicago -Dcompiler.java=11 -Druntime.java=8

```

fails with:

```1> [2018-11-27T09:08:28,186][INFO ][o.e.c.b.CompositeBytesReferenceTests] [testSlice] before test
  1> [2018-11-27T09:08:28,448][INFO ][o.e.c.b.CompositeBytesReferenceTests] [testSlice] after test
  2> REPRODUCE WITH: ./gradlew :server:test -Dtests.seed=24CCFF659D275216 -Dtests.class=org.elasticsearch.common.bytes.CompositeBytesReferenceTests -Dtests.method="testSlice" -Dtests.security.manager=true -Dtests.locale=es-CR -Dtests.timezone=America/Chicago -Dcompiler.java=11 -Druntime.java=8
ERROR   0.35s | CompositeBytesReferenceTests.testSlice <<< FAILURES!
   > Throwable #1: java.lang.ArrayIndexOutOfBoundsException: -1
   >    at __randomizedtesting.SeedInfo.seed([24CCFF659D275216:336770BBDA019DDE]:0)
   >    at org.elasticsearch.common.bytes.CompositeBytesReference.slice(CompositeBytesReference.java:87)
   >    at org.elasticsearch.common.bytes.AbstractBytesReferenceTestCase.testSlice(AbstractBytesReferenceTestCase.java:75)
   >    at java.lang.Thread.run(Thread.java:748)


```

### PR DESCRIPTION
**Pull request by jpountz**
CompositeBytesReference#slice has two bugs:
 - One that makes it fail if the reference is empty and an empty slice is
   created, this is #35950 and is fixed by special-casing empty-slices.
 - One performance bug that makes it always create a composite slice when
   creating a slice that ends on a boundary, this is fixed by computing `limit`
   as the index of the sub reference that holds the last element rather than
   the next element after the slice.

Closes #35950

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/bytes/CompositeBytesReference.java [hunk 1]
```diff
@@ -77,10 +78,16 @@ public int length() {
 
     @Override
     public BytesReference slice(int from, int length) {
+        FutureObjects.checkFromIndexSize(from, length, this.length);
+
+        if (length == 0) {
+            return new BytesArray(BytesRef.EMPTY_BYTES);
```
Comment (tbrooks8) [inline server/src/main/java/org/elasticsearch/common/bytes/CompositeBytesReference.java]: Can you use `BytesArray.EMPTY` here?

Comment (jpountz) [reply server/src/main/java/org/elasticsearch/common/bytes/CompositeBytesReference.java]: Even better, thanks for the pointer.

**Review by Tim-Brooks — APPROVED** [review]
LGTM

**Review by s1monw — APPROVED** [review]
LGTM

**Pull request merged**